### PR TITLE
Document azidentity token caching

### LINKED ISSUE #19358: Create documentation for Identity credential cache
Our current documentation for how our SDK interacts with the MSAL token cache is insufficient. We need to doc:

- What scenarios caching is useful for
- How caching is exposed from our library
- What expectations should be when caching is used (performance, throttling)
- How the disk cache is secured when used

### LINKED ISSUE #19660: Document which credential types have built-in token caching
The Azure Identity module's README doesn't indicate which credential types have built-in token caching. Customers must read each credential type's code to figure it out. This info should be available at [Credential Types](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/README.md#credential-types). Consider adding a "Built-in token caching?" column to each of these tables.

### LINKED ISSUE #21343: Add Section on Clearing Token Cache to TOKEN_CACHING.md
Tracking Issue for Go for the future:

Customers have asked us how to clear the token cache shared by MSAL and Azure Identity.  Currently there is no way to clear the cache, and we should add a section heading with this information (and a sentence on how to avoid using the cache).

Similar to this document here: https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/identity/azure-identity/TOKEN_CACHING.md

### PR DESCRIPTION
**Pull request by chlowell**
This doc, largely copied from other SDKs, describes in detail how credentials cache tokens. Closes #19358, closes #19660, closes #21343

### CODE REVIEW

File: sdk/azidentity/TOKEN_CACHING.MD [hunk 1]
```diff
@@ -0,0 +1,67 @@
+## Token caching in the Azure Identity client library
+
+*Token caching* is a feature provided by the Azure Identity library that allows apps to:
```
Comment (jhendrixMSFT) [inline sdk/azidentity/TOKEN_CACHING.MD:3]: Do we use the term "Azure Identity library" elsewhere or should it be "`azidentity` module"?  Not advocating for either, just as long as it's consistent.

File: sdk/azidentity/TOKEN_CACHING.MD [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+### In-memory token caching
+
+*In-memory token caching* is the default option provided by the Azure Identity library. This caching approach allows apps to store access tokens in memory. With in-memory token caching, the library first determines if a valid access token for the requested resource is already stored in memory. If a valid token is found, it's returned to the app without the need to make another request to Entra ID. If a valid token isn't found, the library will automatically acquire a token by sending a request to Entra ID. The in-memory token cache provided by the Azure Identity library is thread-safe.
+
+**Note:** When Azure Identity library credentials are used with Azure service libraries (for example, Azure Blob Storage), the in-memory token caching is active in the `Pipeline` layer as well. All `TokenCredential` implementations are supported there, including custom implementations external to the Azure Identity library.
+
+#### Caching cannot be disabled
+
+As there are many levels of cache, it's not possible disable in-memory caching. However, the in-memory cache may be cleared by creating a new credential instance.
```
Comment (jhendrixMSFT) [inline sdk/azidentity/TOKEN_CACHING.MD:21]: As there are many levels of caching, it's not possible to disable in-memory caching.

**Review by jhendrixMSFT — APPROVED** [review]

File: sdk/azidentity/TOKEN_CACHING.MD [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+### Credentials supporting token caching
+
+The following table indicates the state of in-memory and persistent caching in each credential type.
+
+**Note:** In-memory caching is activated by default. Persistent token caching needs to be enabled as shown in [this example](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity@v1.5.0-beta.1#example-package-PersistentCache).
+
+| Credential                     | In-memory token caching                                             | Persistent token caching |
+|--------------------------------|---------------------------------------------------------------------|--------------------------|
+| `AzureCLICredential`           | Not Supported                                                       | Not Supported            |
```
Comment (scottaddie) [inline sdk/azidentity/TOKEN_CACHING.MD:56]: Should `EnvironmentCredential` be listed too?

Comment (chlowell) [reply sdk/azidentity/TOKEN_CACHING.MD:56]: Yes, I'll add it

File: sdk/azidentity/TOKEN_CACHING.MD (hunk 1 above)
Comment (chlowell) [reply sdk/azidentity/TOKEN_CACHING.MD:3]: We use "Azure Identity client module" in the readme, so I'll use that here as well

**Review by scottaddie — APPROVED** [review]

**Comment by scottaddie** [comment]
@chlowell Please also add a section to the README. See https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/identity/Azure.Identity#token-caching for an example.

**Pull request merged**